### PR TITLE
Add OSP MachineSet docs - BZ 1822065

### PR DESCRIPTION
**Pull request by maxwelldb**
/wip

Context: https://bugzilla.redhat.com/show_bug.cgi?id=1822065

### CODE REVIEW

**Comment by maxwelldb** [comment]
@pierreprinetti Any thoughts on this? 

1. Is the YAML correct and sufficient to get a user started with this task?
2. Is the YAML as it's situated in its assembly okay (let me know if you'd like me to generate a preview)?

File: machine_management/creating_machinesets/creating-machineset-osp.adoc [hunk 1]
```diff
@@ -0,0 +1,17 @@
+[id="creating-machineset-osp"]
+= Creating a MachineSet on OpenStack
+include::modules/common-attributes.adoc[]
+:context: creating-machineset-osp
+
+You can create a different MachineSet to serve a specific purpose in your
+{product-title} cluster on {rh-openstack-first}. For example, you might
+create infrastructure MachineSets and related Machines so that you can move
+supporting workloads to the new Machines.git st
```
Comment (pierreprinetti) [inline machine_management/creating_machinesets/creating-machineset-osp.adoc:9]: git st?

File: modules/machineset-yaml-osp.adoc [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        value:
+          apiVersion: openstackproviderconfig.openshift.io/v1alpha1
+          cloudsSecret:
+            name: openstack-cloud-credentials
+            namespace: openshift-machine-api
+          flavor: NOVA_FLAVOR
+          image: GLANCE_IMAGE_NAME_OR_LOCATION
+          internalLoadBalancer: ""
+          kind: OpenstackProviderSpec
+          location: centralus
```
Comment (pierreprinetti) [inline modules/machineset-yaml-osp.adoc:52]: where is this coming from?

File: modules/machineset-yaml-osp.adoc [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+[source,yaml]
+----
+apiVersion: machine.openshift.io/v1beta1
+kind: MachineSet
+metadata:
+  labels:
+    machine.openshift.io/cluster-api-cluster: INFRASTRUCTURE_ID <1>
+    machine.openshift.io/cluster-api-machine-role: NODE_ROLE <2>
+    machine.openshift.io/cluster-api-machine-type: NODE_ROLE <2>
+  name: INFRASTRUCTURE_ID-NODE_ROLE-REGION <3>
```
Comment (pierreprinetti) [inline modules/machineset-yaml-osp.adoc:26]: `region` is [currently](https://github.com/openshift/installer/blob/189b16eba4e0d3769bc1179181311fbb0ed92453/pkg/asset/machines/openstack/machinesets.go#L39) not in the machineset.

Comment (pierreprinetti) [reply modules/machineset-yaml-osp.adoc:26]: not in the OpenStack machineset anyway

File: modules/machineset-yaml-osp.adoc (hunk 2 above)
Comment (maxwelldb) [reply modules/machineset-yaml-osp.adoc:52]: Unclear. Probably from reusing parts of existing MachineSet docs samples. Removing "internalLoadBalancer" and "location."

File: modules/machineset-yaml-osp.adoc (hunk 3 above)
Comment (maxwelldb) [reply modules/machineset-yaml-osp.adoc:26]: And removing references to `region`.

File: modules/machineset-yaml-osp.adoc (hunk 2 above)
Comment (maxwelldb) [reply modules/machineset-yaml-osp.adoc]: This now matches the example in the BZ ticket, as far as I can tell.

File: modules/machineset-yaml-osp.adoc [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+      labels:
+        machine.openshift.io/cluster-api-cluster: INFRASTRUCTURE_ID <1>
+        machine.openshift.io/cluster-api-machine-role: NODE_ROLE <2>
+        machine.openshift.io/cluster-api-machine-type: NODE_ROLE <2>
+        machine.openshift.io/cluster-api-machineset: INFRASTRUCTURE_ID-NODE_ROLE <3>
+    spec:
+      providerSpec:
+        value:
+          apiVersion: openstackproviderconfig.openshift.io/v1alpha1
+          cloudsSecret:
```
Comment (pierreprinetti) [inline modules/machineset-yaml-osp.adoc:45]: There's also `cloudName: openstack`

Comment (maxwelldb) [reply modules/machineset-yaml-osp.adoc:45]: Thanks!

**Comment by maxwelldb** [comment]
With corrections: https://osp-machineset-1822065--ocpdocs.netlify.app/openshift-enterprise/latest/machine_management/creating_machinesets/creating-machineset-osp.html

**Comment by pierreprinetti** [comment]
LGTM :+1:

**Review by pierreprinetti — APPROVED** [review]

File: modules/machineset-yaml-osp.adoc [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+          - filter: {}
+            name: <infrastructure_ID>-<node_role>
+          serverMetadata:
+            Name: <infrastructure_ID>-<node_role>
+            openshiftClusterID: <infrastructure_ID>
+          tags:
+          - openshiftClusterID=<infrastructure_ID>
+          trunk: true
+          userDataSecret:
+            name: <node_role>-user-data <2>
```
Comment (luis5tb) [inline modules/machineset-yaml-osp.adoc:68]: this is missing information about how to create this secret (that needs to point to the proper machineconfigpool (for instance kubelet will need to be changed to start with the new role too

Comment (maxwelldb) [reply modules/machineset-yaml-osp.adoc:68]: @luis5tb Thanks for looking! 

This is a problem common to all of the machineSet docs. Are there OSP-specific parts to this procedure?

File: modules/machineset-yaml-osp.adoc [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+          kind: OpenstackProviderSpec
+          networks:
+          - filter: {}
+            subnets:
+            - filter:
+                name: <subnet_name>
+                tags: openshiftClusterID=<infrastructure_ID>
+          securityGroups:
+          - filter: {}
+            name: <infrastructure_ID>-<node_role>
```
Comment (yrro) [inline modules/machineset-yaml-osp.adoc:60]: Should be `<infrastructure_ID>-worker` to re-use the worker security group.

The instructions for AWS for instance re-use the worker security group.

File: modules/machineset-yaml-osp.adoc (hunk 5 above)
Comment (yrro) [reply modules/machineset-yaml-osp.adoc:68]: The AWS and GCP instructions re-use the `worker-user-data` secret. The Azure instructions use `<role>-user-data` but similarly do not explain how to create it.

**Comment by maxwelldb** [comment]
> The AWS and GCP instructions re-use the `worker-user-data` secret. The Azure instructions use `<role>-user-data` but similarly do not explain how to create it. 

@yrro @luis5tb What's the best way to accomplish this? Search didn't return anything in our docs set about it.

File: modules/machineset-yaml-osp.adoc (hunk 6 above)
Comment (maxwelldb) [reply modules/machineset-yaml-osp.adoc:60]: Thanks!

**Comment by yrro** [comment]
> > The AWS and GCP instructions re-use the `worker-user-data` secret. The Azure instructions use `<role>-user-data` but similarly do not explain how to create it.
> 
> @yrro @luis5tb What's the best way to accomplish this? Search didn't return anything in our docs set about it.

I exported `worker-user-data` the secret, changed the URL in `.ignition.config.append.source` URL to retrieve `/config/infra` rather than `/config/worker` and used it to create the `infra-user-data` secret.

(I previously created an `infra` MachineConfigPool as described in https://access.redhat.com/solutions/4342791, IMO that should be folded into this section of the docs as well, but there's another issue for that: #20688.

**Comment by maxwelldb** [comment]
My inclination is to get this to parity with the other docs and fix any common gaps w/another bug + PR, including what's suggested in #20688. Any thoughts? @luis5tb @yrro

**Comment by yrro** [comment]
Sure. For reference I've spotted these inconsistencies:

* Azure docs set `machine.openshift.io/cluster-api-machine-role` and `machine.openshit.io/cluster-api-machine-type` labels; AWS/GCP do not.
* [nit] Azure/GBP docs set `creationTimestamp: null`; AWS does not.
* Azure refers to `infra-user-data` secret (implying the creation of an `<infrastructureID>-infra` MachineConfigPool; AWS/GCP re-used `worker-user-data` (hence re-using the `<infrastructureID>-worker` MachineConfigPool).

I would err on the side of re-using the `worker-user-secret` userData secret & hence the `<infrastructureID>-worker MachineConfigPool`, on the grounds that setting up a new `MachineConfigPool` & userData secret is not documented elsewhere, and adds extra un-necessary steps to the task of creating an `infra` MachineSet, which is a very common Day 2 task.

**Comment by maxwelldb** [comment]
@yrro Way helpful comment. Thanks!

File: modules/machineset-yaml-osp.adoc (hunk 5 above)
Comment (maxwelldb) [reply modules/machineset-yaml-osp.adoc:68]: Deferring as per https://github.com/openshift/openshift-docs/pull/21896#issuecomment-641298033

**Comment by maxwelldb** [comment]
@openshift/team-documentation PTAL at this. No rush.

**Comment by maxwelldb** [comment]
@sunzhaohua2 Thanks!

**Comment by maxwelldb** [comment]
Easy preview link: https://osp-machineset-1822065--ocpdocs.netlify.app/openshift-enterprise/latest/machine_management/creating_machinesets/creating-machineset-osp.html

**Review by kalexand-rh — COMMENTED** [review]
LGTM

File: modules/machineset-yaml-osp.adoc (hunk 6 above)
Comment (maxwelldb) [reply modules/machineset-yaml-osp.adoc:60]: Deferring (https://github.com/openshift/openshift-docs/pull/21896#issuecomment-641298033)

**Pull request merged**